frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+
+		if val_existed && !val_exists {
+			// Value was deleted
+			Counter::mutate(|value| value.saturating_dec());
+		} else if !val_existed && val_exists {
+			// Value was added
+			Counter::mutate(|value| value.saturating_inc());
+		}
+
+		res
+	}
+
+	/// Get the storage key used to fetch a value corresponding to a specific key.
+	pub fn hashed_key_for<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Vec<u8> {
+		Map::hashed_key_for(key)
+	}
+
+	/// Does the value (explicitly) exist in storage?
+	pub fn contains_key<KeyArg: EncodeLike<Key>>(key: KeyArg) -> bool {
+		Map::contains_key(key)
+	}
+
+	/// Load the value associated with the given key from the map.
+	pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Map::Query {
+		Map::get(key)
+	}
+
+	/// Try to get the value for the given key from the map.
+	///
+	/// Returns `Ok` if it exists, `Err` if not.
+	pub fn try_get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Result<Value, ()> {
+		Map::try_get(key)
+	}
+
+	/// Swap the values of two keys.
+	pub fn swap<KeyArg1: EncodeLike<Key>, KeyArg2: EncodeLike<Key>>(key1: KeyArg1, key2: KeyArg2) {
+		Map::swap(key1, key2)
+	}
+
+	/// Store a value to be associated with the given key from the map.
+	pub fn insert<KeyArg: EncodeLike<Key> + Clone, ValArg: EncodeLike<Value>>(key: KeyArg, val: ValArg) {
+		if !Map::contains_key(key.clone()) {
+			Counter::mutate(|value| value.saturating_inc());
+		}
+		Map::insert(key, val)
+	}
+
+	/// Remove the value under a key.
+	pub fn remove<KeyArg: EncodeLike<Key> + Clone>(key: KeyArg) {
+		if Map::contains_key(key.clone()) {
+			Counter::mutate(|value| value.saturating_dec());
+		}
+		Map::remove(key)
+	}
+
+	/// Mutate the value under a key.
+	pub fn mutate<KeyArg: EncodeLike<Key> + Clone, R, F: FnOnce(&mut Map::Query) -> R>(
+		key: KeyArg,
+		f: F
+	) -> R {
+		Self::mutate_counter(
+			Map::mutate,
+			key,
+			f,
+		)
+	}
+
+	/// Mutate the item, only if an `Ok` value is returned.
+	pub fn try_mutate<KeyArg, R, E, F>(key: KeyArg, f: F) -> Result<R, E>
+	where
+		KeyArg: EncodeLike<Key> + Clone,
+		F: FnOnce(&mut Map::Query) -> Result<R, E>,
+	{
+		Self::mutate_counter(
+			Map::try_mutate,
+			key,
+			f,
+		)
+	}
+
+	/// Mutate the value under a key. Deletes the item if mutated to a `None`.
+	pub fn mutate_exists<KeyArg: EncodeLike<Key> + Clone, R, F: FnOnce(&mut Option<Value>) -> R>(
+		key: KeyArg,
+		f: F
+	) -> R {
+		Self::mutate_counter(
+			Map::mutate_exists,
+			key,
+			f,
+		)
+	}
+
+	/// Mutate the item, only if an `Ok` value is returned. Deletes the item if mutated to a `None`.
+	pub fn try_mutate_exists<KeyArg, R, E, F>(key: KeyArg, f: F) -> Result<R, E>
+	where
+		KeyArg: EncodeLike<Key> + Clone,
+		F: FnOnce(&mut Option<Value>) -> Result<R, E>,
+	{
+		Self::mutate_counter(
+			Map::try_mutate_exists,
+			key,
+			f,
+		)
+	}
+
+	/// Take the value under a key.
+	pub fn take<KeyArg: EncodeLike<Key> + Clone>(key: KeyArg) -> Map::Query {
+		if Map::contains_key(key.clone()) {
```

Comment:
is `Map::Query` the Option type? if yes you can just check if it is None

Reply:
I cannot assume it is the option type.

The problem is that the Query can either return the default value or be an option, and when it returns the default value, it is not guaranteed that it was not in storage, set as the default value or not existing at all.

I think the only thing you can do here is actually call the `contains_key` function which inspects exactly what we want.

Reply:
in that case you can always use mutate_exists

Reply:
yes take is now optimized using mutate_exists